docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1844,16 +1855,100 @@ entry value. When providing serializable enum values through `P4Data`, one can
+to `typedef`, this mechanism introduces in fact a new type, which is not a
+strict synonym of the original type. It is important to preserve this
+distinction in the P4Info message, in particular for the purposes of
+[translation](#sec-psa-metadata-translation). When introducing a new type, the
+declaration can be annotated with `@p4runtime_translation` to indicate that the
+type exposed to the P4Runtime client is different from the original P4 type. One
+important use-case is for [port numbers](#sec-translation-of-port-numbers),
+whose underlying dataplane representation may vary on different targets, but for
+which it may be convenient to present a unified representation and numbering
+scheme to the control-plane. **The `@p4runtime_translation` annotation can only
+be used if the underlying P4 built-in type is a fixed-width unsigned bitstring
+type (`bit<W>`)** and the type exposed to the control-plane will also be a
+fixed-width unsigned bitstring, with a potentially different bitwidth. It takes
+two parameters: a *URI* (Uniform Resource Identifier) which uniquely identifies
+the translation being performed on entities of the new type to the P4Runtime
+server and the *bitwidth* of the bitstring type exposed to the control-plane. It
+is recommended that the URI includes at least the P4 architecture name and the
+type name.
+
+User-defined types are specified using the `P4NewTypeSpec` message, which has
+the following fields:
+
+* `representation`, a Protobuf `oneof` specifying how values of this type are
+  exchanged between client and server; it can be either:
+
+    * `original_type`, if and only if no `@p4runtime_translation` annotation is
+      present. It specifies the underlying built-in P4 type for the user-defined
+      type. If the underlying type used in the P4 `type` declaration is itself a
+      user-defined type, `original_type` is obtained by "walking" the chain of
+      `type` declarations recursively until a built-in type (e.g `bit<W>` is
+      found).
+
+    * `translated_type`, if and only if the P4 `type` declaration was annotated
+      with `@p4runtime_translation`. It is of type `P4NewTypeTranslation`, which
+      itself has two fields - `uri` and `sdn_bitwidth`, which map to the two
+      input parameters to the annotation.
+
+* `annotations`, a repeated field of strings, each one representing a P4
+  annotation associated to the type declaration.
+
+For example, an architecture - in this case PSA - may introduce a new type for
+port numbers:
+~ Begin P4Example
+@p4runtime_translation("p4.org/psa/v1/PortId_t", 32)
+type bit<9> PortId_t;
+~ End P4Example
+
+In this case, the P4Info message would include the following `P4TypeInfo`
+message:
+
+~ Begin Prototext
+type_info {
+  new_types {
+    key: "PortId_t"
+    value {  # P4NewTypeSpec
+      original_type {  # P4DataTypeSpec
+        type_spec {
+          bit {
+            bitwidth: 9
+          }
+        }
+      }
```

Comment:
Sorry if I missed this in my first review, but if the goal is to have P4Info contents be identical for all P4 target devices, isn't the presence of `original_type` going to conflict with that goal?

Reply:
Actually, doesn't this example violate the comments mentioned later that at most one of original_type and translated_type should be present?  Should "translated" below be "translated_type"?

Reply:
My bad. I updated the text and the .proto, but not the example.